### PR TITLE
Adjust coingecko prices based on token decimals

### PR DESCRIPTION
**Pull request by MartinquaXD**
# Description
Fixes: https://github.com/cowprotocol/services/issues/2887

# Changes
Adjust returned native token prices when the quoted token has a different amount of decimals than the native token to be inline with all other native price estimator implementations.

Additionally I also removed the option to quote without denominating in any token mainly to reduce complexity.

## How to test
Added an `ignored` unit test that checks that prices make sense.
Also compared prices returned for `wxdai` and `usdc` with prices logged in our infra.

### CODE REVIEW

File: crates/shared/src/price_estimation/native/coingecko.rs [hunk 1]
```diff
@@ -158,19 +178,35 @@ impl CoinGecko {
...
             .cloned()
             .ok_or(PriceEstimationError::NoLiquidity)?
             .try_into()
             .context("failed to convert price to decimal")?;
 
-        Ok(token_price_eth
+        // When the quoted token and the denominator have different number of decimals
+        // the computed price effectively needs to be shifted by the difference.
+        let adjustment =
+            Decimal::new(10, 0).powi(denominator.decimals as i64 - token_decimals as i64);
```
Comment (m-lord-renkse) [inline crates/shared/src/price_estimation/native/coingecko.rs:199]: it could be rare, but 10 ^ 255 would result in panic, we could use `checked_powi` just in case

File: crates/shared/src/price_estimation/native/coingecko.rs [hunk 2]
```diff
@@ -379,4 +438,50 @@ mod tests {
...
+    #[ignore]
+    async fn prices_adjusted_for_token_decimals() {
+        let usdc = addr!("2a22f9c3b484c3629090FeED35F17Ff8F88f76F0");
+        let wxdai = addr!("e91D153E0b41518A2Ce8Dd3D7944Fa863463a97d");
+        let mut mock = MockTokenInfoFetching::new();
+        mock.expect_get_token_infos().returning(move |tokens| {
+            tokens
+                .iter()
+                .map(|t| {
+                    let decimals = if *t == usdc { Some(6) } else { Some(18) };
```
Comment (m-lord-renkse) [inline crates/shared/src/price_estimation/native/coingecko.rs:452]: is it possible that a token has more than 18 decimals? if yes, could you please add a token like that in that test to check that the reverse operation (powi with a negative number + then multiplication to normalize it) works as expected? :pray:

**Review by m-lord-renkse — APPROVED** [review]

File: crates/shared/src/price_estimation/native/coingecko.rs (hunk 1 above)
Comment (m-lord-renkse) [reply crates/shared/src/price_estimation/native/coingecko.rs:199]: What's more, please can you do the casting like `i64::from()` instead of `as`. The reason is that if we one day change the underlying type (`u8`) for whatever reason, we guarantee it will always have a safe casting (because if it doesn't exist the `From<>` trait then it will not compile).

File: crates/shared/src/price_estimation/native/coingecko.rs [hunk 3]
```diff
@@ -127,14 +136,19 @@ impl CoinGecko {
...
     ) -> Result<HashMap<H160, NativePriceEstimateResult>, PriceEstimationError> {
-        tokens.insert(denominator);
-        let prices_in_eth = self.bulk_fetch_denominated_in_eth(&tokens).await?;
+        tokens.insert(self.denominator.address);
+
+        let tokens_vec: Vec<_> = tokens.iter().cloned().collect();
+
+        let (prices_in_eth, infos) = tokio::try_join!(
+            self.bulk_fetch_denominated_in_eth(&tokens),
+            self.infos.get_token_infos(&tokens_vec).map(Result::Ok),
```
[the remark this answers is not in the document]
Comment (MartinquaXD) [reply crates/shared/src/price_estimation/native/coingecko.rs:146]: Technically a good idea but that doesn't play nice with `automock` and `async_trait` unfortunately. :/

File: crates/shared/src/price_estimation/native/coingecko.rs (hunk 2 above)
Comment (MartinquaXD) [reply crates/shared/src/price_estimation/native/coingecko.rs:453]: Yeah, good idea. This was actually a problem in the original fix for the oneinch estimator adjustment since we used `u64` operations there and `10 ** 21` overflowed that value which caused other issues.

File: crates/shared/src/price_estimation/native/coingecko.rs [hunk 4]
```diff
@@ -127,14 +136,19 @@ impl CoinGecko {
     async fn bulk_fetch_denominated_in_token(
         &self,
         mut tokens: HashSet<Token>,
-        denominator: Token,
     ) -> Result<HashMap<H160, NativePriceEstimateResult>, PriceEstimationError> {
-        tokens.insert(denominator);
-        let prices_in_eth = self.bulk_fetch_denominated_in_eth(&tokens).await?;
+        tokens.insert(self.denominator.address);
```
Comment (fleupold) [inline crates/shared/src/price_estimation/native/coingecko.rs:140]: Are we deliberately removing the optimisation of not requesting the price for weth if not needed?

File: crates/shared/src/price_estimation/native/coingecko.rs [hunk 5]
```diff
@@ -158,19 +178,36 @@ impl CoinGecko {
         token: &Token,
         denominator_price_eth: Decimal,
         prices: &HashMap<Token, f64>,
+        infos: &HashMap<Token, TokenInfo>,
```
Comment (fleupold) [inline crates/shared/src/price_estimation/native/coingecko.rs:181]: This method signature is quite 🤡 in my opinion (it has 5 values of which some are redundant).

I think we should simply pass it the token as `(TokenInfo, priceInEth)` and the denominator as `(Denominator, priceInEth)` and keep this method for the pure non-trivial math.

File: crates/shared/src/price_estimation/native/coingecko.rs [hunk 6]
```diff
@@ -158,19 +178,36 @@ impl CoinGecko {
...
+        // When the quoted token and the denominator have different number of decimals
+        // the computed price effectively needs to be shifted by the difference.
+        let adjustment = Decimal::new(10, 0)
+            .checked_powi(i64::from(denominator.decimals) - i64::from(token_decimals))
+            .context("price adjustment overflows Decimal")?;
+
+        let price_denominated_in_token = token_price_eth
             .checked_div(denominator_price_eth)
             .context("division by zero")?
+            .checked_mul(adjustment)
```
Comment (fleupold) [inline crates/shared/src/price_estimation/native/coingecko.rs:205]: Do we need to do the adjustment before or after the division to not lose precision?

File: crates/shared/src/price_estimation/native/coingecko.rs (hunk 5 above)
Comment (MartinquaXD) [reply crates/shared/src/price_estimation/native/coingecko.rs:181]: The alternative to this function seems to be even more 🤡 error handling in the calling function. While your suggestion would make the conversion more pure it would make the surrounding code way harder to read.
Unless I'm missing some trick to reasonably manage errors (using anonymous closures to get early returns doesn't count IMO).

File: crates/shared/src/price_estimation/native/coingecko.rs (hunk 6 above)
Comment (MartinquaXD) [reply crates/shared/src/price_estimation/native/coingecko.rs:205]: I don't think so. Since this is not using a float under the hood but a `(base: i64, scale: u32)` the order of operations should not matter AFAIU.

File: crates/shared/src/price_estimation/native/coingecko.rs (hunk 4 above)
Comment (MartinquaXD) [reply crates/shared/src/price_estimation/native/coingecko.rs:140]: Yes, this was a conscious decision for the sake of simplicity. I'm happy to put the optimization back in if we actually need the extra 5% of batch size but for now I would prefer to keep the code leaner since this file is already on the more complex side IMO.

**Review by squadgazzz — APPROVED** [review]
LGTM

File: crates/shared/src/price_estimation/native/coingecko.rs (hunk 5 above)
Comment (fleupold) [reply crates/shared/src/price_estimation/native/coingecko.rs:181]: I mean the fact that you are passing in `denominator_price_eth` on top of prices is just redundant (and seems inconsistent) so at least you could either
- remove denominator_price_eth and and "expect" on it being in prices 
- pass the token price into the function directly

But I also don't really see how error handling is an issue here (you are bubbling up errors in any case)

**Review by fleupold — APPROVED** [review]
Nothin blocking, and we should probably try to get this into the release so that coingecko can have some impact.
